dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java

```diff
@@ -37,177 +37,225 @@
+        fieldFilterService.toObjectNodes(icons.stream().toList(), iconRequestParams.getFields());
+
+    linkService.generatePagerLinks(pager, CustomIcon.class);
+
+    return new PaginatedIconResponse(pager, objectNodes);
   }
 
   @GetMapping(value = "/{key}/icon")
   public void getIconData(@PathVariable String key, HttpServletResponse response)
       throws NotFoundException, WebMessageException, IOException {
 
-    Icon icon = iconService.getIcon(key);
+    CustomIcon customIcon = iconService.getCustomIcon(key);
 
-    if (icon instanceof DefaultIcon) {
-      downloadDefaultIcon(icon.getKey(), response);
-    } else if (icon instanceof CustomIcon customIcon) {
-      downloadCustomIcon(customIcon.getFileResourceUid(), response);
+    if (!customIcon.getCustom()) {
+      downloadDefaultIcon(customIcon.getKey(), response);
+    } else {
+      downloadCustomIcon(customIcon.getFileResource(), response);
     }
   }
 
-  @GetMapping
-  public @ResponseBody PaginatedIconResponse getAllIcons(IconRequestParams iconRequestParams)
-      throws BadRequestException {
-
-    IconOperationParams iconOperationParams = iconRequestParamsMapper.map(iconRequestParams);
+  @GetMapping(value = "/{key}")
+  public ResponseEntity<CustomIcon> getIconByKey(@PathVariable String key)
+      throws NotFoundException, WebMessageException {
 
-    Pager pager = null;
+    CustomIcon customIcon = iconService.getCustomIcon(key);
 
-    if (iconRequestParams.isPaging()) {
-      long total = iconService.count(iconOperationParams);
-      pager = new Pager(iconRequestParams.getPage(), total, iconRequestParams.getPageSize());
-      iconOperationParams.setPager(pager);
+    if (customIcon == null) {
+      throw new WebMessageException(
+          WebMessageUtils.notFound(String.format("CustomIcon with key %s not found", key)));
     }
 
-    List<IconResponse> iconResponses =
-        iconService.getIcons(iconOperationParams).stream().map(iconMapper::from).toList();
+    customIcon.setHref(
+        customIcon.getCustom() ? getCustomIconReference(key) : getDefaultIconReference(key));
 
-    List<ObjectNode> objectNodes =
-        fieldFilterService.toObjectNodes(iconResponses, iconRequestParams.getFields());
+    return new ResponseEntity<>(customIcon, HttpStatus.OK);
+  }
 
-    linkService.generatePagerLinks(pager, IconResponse.class);
+  @GetMapping("/{key}/icon.svg")
+  @Deprecated
+  public void getIconData(
+      HttpServletResponse response, HttpServletRequest request, @PathVariable String key)
+      throws IOException, WebMessageException {
 
-    return new PaginatedIconResponse(pager, objectNodes);
+    if (!iconService.customIconExists(key)) {
+      throw new WebMessageException(
+          WebMessageUtils.notFound(String.format("CustomIcon with key %s not found", key)));
+    }
+
+    String location = response.encodeRedirectURL("/icons/" + key + "/icon");
+    response.sendRedirect(ContextUtils.getRootPath(request) + location);
```

Comment:
Why do we redirect?

Reply:
https://dhis2.atlassian.net/browse/DHIS2-16041